modules/arcore/arcore_interface.cpp

```diff
@@ -0,0 +1,649 @@
+
+		// === ATTENTION!  ATTENTION!  ATTENTION! ===
+		// This method can and will fail in user-facing situations.  Your
+		// application must handle these cases at least somewhat gracefully.  See
+		// HelloAR Java sample code for reasonable behavior.
+		if (ArSession_create(env, context, &ar_session) != AR_SUCCESS || ar_session == NULL) {
+			print_line("Godot ARCore: ARCore couldn't be created.");
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+
+		print_line("Godot ARCore: Create ArFrame.");
+
+		ArFrame_create(ar_session, &ar_frame);
+		if (ar_frame == NULL) {
+			print_line("Godot ARCore: Frame couldn't be created.");
+
+			ArSession_destroy(ar_session);
+			ar_session = NULL;
+
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+
+		// Get our size, make sure we have these in portrait
+		Size2 size = OS::get_singleton()->get_window_size();
+		if (size.x > size.y) {
+			width = size.y;
+			height = size.x;
+		} else {
+			width = size.x;
+			height = size.y;
+		}
+
+		// Trigger display rotation
+		display_rotation = -1;
+
+		print_line("Godot ARCore: Resumed.");
+		init_status = INITIALISED;
+	}
+
+	if (init_status == INITIALISED && ar_session != NULL) {
+		ArStatus status = ArSession_resume(ar_session);
+		if (status != AR_SUCCESS) {
+			// uh, what do we do in this situation?
+		}
+	}
+}
+
+void ARCoreInterface::_pause() {
+	if (ar_session != NULL) {
+		ArSession_pause(ar_session);
+	}
+}
+
+void ARCoreInterface::notification(int p_what) {
+	// Needs testing, this should now be called
+
+	ARVRServer *arvr_server = ARVRServer::get_singleton();
+	ERR_FAIL_NULL(arvr_server);
+
+	switch (p_what) {
+		case MainLoop::NOTIFICATION_WM_FOCUS_IN: {
+			if (is_initialized()) {
+				_resume();
+
+				if (init_status == INITIALISE_FAILED) {
+					arvr_server->clear_primary_interface_if(this);
+				}
+			}
+		}; break;
+		case MainLoop::NOTIFICATION_WM_FOCUS_OUT:
+			if (is_initialized()) {
+				_pause();
+			}
+			break;
+		default:
+			break;
+	}
+}
+
+bool ARCoreInterface::initialize() {
+	// TODO we may want to check for status PAUZED and just call resume (if we decide to implement that)
+
+	ARVRServer *arvr_server = ARVRServer::get_singleton();
+	ERR_FAIL_NULL_V(arvr_server, false);
+
+	if (init_status == INITIALISE_FAILED) {
+		// if we fully failed last time, don't try again..
+		return false;
+	} else if (init_status == NOT_INITIALISED) {
+		print_line("Godot ARCore: Initialising...");
+		init_status = START_INITIALISE;
+
+		// create our camera feed
+		if (feed.is_null()) {
+			print_line("Godot ARCore: Creating camera feed...");
+
+			feed.instance();
+			feed->set_name("ARCore");
+			feed->set_active(true);
+
+			CameraServer *cs = CameraServer::get_singleton();
+			if (cs != NULL) {
+				cs->add_feed(feed);
+
+				print_line("Godot ARCore: Feed " + itos(feed->get_id()) + " added");
+			}
+		}
+
+		// call resume for the first time ;)
+		_resume();
```

Comment:
Seems a bit janky to also be invoking `_resume()` here. It might make more sense to have the init logic in `_resume()` here instead.

Reply:
If I recall correctly the problem is the scenario where you haven't given permission yet because you are running this as your very first ARCore app. Resume is called here but fails, then the user answers the camera permission question and/or installs ARCore and focus returns to our app. 
Resume is then called from NOTIFICATION_WM_FOCUS_IN/NOTIFICATION_APP_RESUMED and now the init succeeds. 

If we put the init in the code here it will fail the first time the app is run, and the user will have to close and re-open the app.

Reply:
The `ARCoreSetupActivity` ensures that all the required permissions and/or installs are done before this code is run, so the logic here can assume that by the time it's invoked, everything is good to go!